File: Lab 4 (13 February 2025)/Convex hull naive approach.cpp
#include <iostream>
#include <vector>
using namespace std;

struct Point {
    int x, y;
};

int orientation(Point a, Point b, Point c)
{
    int val = (b.y - a.y)*(c.x - b.x) - (b.x - a.x)*(c.y - b.y);

    if (val == 0) return 0;
    return (val > 0) ? 1 : -1;
}

void convexHullNaive(vector<Point> &points)
{
    int n = points.size();
    if (n < 3)
    {
        cout << "Convex hull not possible\n";
        return;
    }
    cout << "Convex Hull points:\n";
    for (int i = 0; i < n; i++)
    {
        for (int j = i+1; j < n; j++)
        {
            int pos = 0, neg = 0;
            for (int k = 0; k < n; k++)
            {
                if (k == i || k == j) continue;
                int dir = orientation(points[i], points[j], points[k]);
                if (dir > 0) pos++;
                else if (dir < 0) neg++;
            }
            if (pos == 0 || neg == 0)
            {
                cout << "(" << points[i].x << ", " << points[i].y << ") ";
                cout << "(" << points[j].x << ", " << points[j].y << ")\n";
            }
        }
    }
}

int main()
{
    vector<Point> points = {{0, 3}, {2, 2}, {1, 1}, {2, 1}, {3, 0}, {0, 0}, {3, 3}};
    convexHullNaive(points);
    return 0;
}

File: Lab 4 (13 February 2025)/Convex hull naive approach.py
import time
import random
import matplotlib.pyplot as p

def is_left(p, q, r):
    return (q[0] - p[0]) * (r[1] - p[1]) - (q[1] - p[1]) * (r[0] - p[0]) >= 0

def convex_hull_naive(points):
    n = len(points)
    hull = set()
    for i in range(n):
        for j in range(i+1, n):
            left = right = 0
            for k in range(n):
                if k == i or k == j:
                    continue
                if is_left(points[i], points[j], points[k]):
                    left += 1
                else:
                    right += 1
            if left == 0 or right == 0:
                hull.add(points[i])
                hull.add(points[j])
    return list(hull)

def generate_points(n):
    return [(random.randint(0, 1000), random.randint(0, 1000)) for _ in range(n)]

sizes = list(range(10, 111, 10))
times = []

for n in sizes:
    pts = generate_points(n)
    start = time.time()
    convex_hull_naive(pts)
    end = time.time()
    times.append(end - start)

p.plot(sizes, times, marker='o', label='Naive Convex Hull (O(n³))')
p.xlabel('Number of Points')
p.ylabel('Execution Time (seconds)')
p.title('Time Complexity of Naive Convex Hull')
p.grid(True)
p.legend()
p.show()